### PR TITLE
Introduce Assignment deadlines 

### PR DESCRIPTION
**Pull request by nwoodthorpe**
~~[WIP - Not everything discussed below has been implemented yet]~~

This PR introduces a V1 assignment deadline functionality for Classroom. The initial implementation will be fairly minimal and lightweight, we plan on building on top of this in the future.

When creating an assignment, teachers will be able to select a deadline (date/time). At this time, Classroom will fetch the latest commit SHA from each assignment repo and save that as the "submission SHA". Teachers will be able to view commits that happened before and after the deadline in a clear and simple way.

Teachers will also be able to modify or remove the deadline via assignment settings. 

One thing to note: if a student force-pushes over their submission commit, all commits will be marked as being after the deadline. In the future, we may explore solutions to that problem such as protecting the default branch, or removing access from the student.

We'll also be exploring better submission evaluation methods after this PR goes through.

I'll upload some screenshots once I've done more of the front-end stuff.

If anyone has feedback on how this is being implemented, let us know :)

Relevant people:

@johndbritton @tarebyte @mozzadrella @sanicki @iamgoodbytes @ppannuto @samwhitlock @arjenpdevries @dhowe @M4urice

### CODE REVIEW

**Comment by nwoodthorpe** [comment]
The deadline can be viewed on assignment/show here:
screen shot 2017-05-30 at 4 58 47 pm

I think as deadlines get more complex, they'll be given their own section. For now, its just the one value though, so I put it beside the assignment type.

In assignment#new/edit, we can set/edit the deadline.
screen shot 2017-05-30 at 5 01 11 pm

Clicking the field opens this dialog:
screen shot 2017-05-30 at 5 02 02 pm

Emptying the field and updating the assignment will delete the deadline.

**Comment by nwoodthorpe** [comment]
Removed the dependency that was causing license issues. That change makes it really hard to display the deadline in local time, so I've decided to use a relative time instead. The exact deadline can still be viewed in assignment settings. If we think it's important enough to have the exact deadline here, it can happen, it just means more messy timezone javascript.

I love timezones

screen shot 2017-05-31 at 4 11 10 pm

**Comment by nwoodthorpe** [comment]
After a deadline, this is what teacher's will see:



Clicking the green icon will bring the teacher to the students repo at the time of submission.

**Comment by nwoodthorpe** [comment]
Looks like the hook above failed: https://travis-ci.org/education/classroom/builds/239663538

This is 100% working now, just need to do a bit work around how to introduce the new feature to users.

**Comment by nwoodthorpe** [comment]
Review @johndbritton @tarebyte when you have time. It's huge, but most of it is specs/cassettes

**Comment by nwoodthorpe** [comment]
screen shot 2017-06-05 at 3 40 05 pm

Any opinions on the text I've chosen for underneath the deadlines box?

File: Gemfile.lock [hunk 1]
```diff
@@ -505,4 +507,4 @@ RUBY VERSION
    ruby 2.4.1p111
 
 BUNDLED WITH
-   1.15.1
+   1.15.0
```
Comment (nwoodthorpe) [inline Gemfile.lock]: ^ Dont do this

File: app/controllers/assignments_controller.rb [hunk 2]
```diff
@@ -12,6 +12,8 @@ def new
 
   def create
     @assignment = Assignment.new(new_assignment_params)
+
+    @assignment.deadline = Deadline.build_from_string(deadline_at: params[:assignment][:deadline]) if deadlines_enabled?
```
Comment (nwoodthorpe) [inline app/controllers/assignments_controller.rb]: Pull into factory module

File: app/controllers/assignments_controller.rb [hunk 3]
```diff
@@ -12,6 +12,8 @@ def new
 
   def create
     @assignment = Assignment.new(new_assignment_params)
```
Comment (nwoodthorpe) [inline app/controllers/assignments_controller.rb]: Add deadline to new_assignment_params

File: app/controllers/assignments_controller.rb [hunk 4]
```diff
@@ -98,4 +100,8 @@ def student_identifier_type_params
       .require(:student_identifier_type)
       .permit(:id)
   end
+
+  def deadlines_enabled?
```
Comment (nwoodthorpe) [inline app/controllers/assignments_controller.rb]: move this to feature_flag_dependency module

File: app/controllers/assignments_controller.rb [hunk 5]
```diff
@@ -98,4 +100,8 @@ def student_identifier_type_params
...
+    current_user.feature_enabled?(:deadlines)
```
Comment (nwoodthorpe) [inline app/controllers/assignments_controller.rb]: &. here

File: app/helpers/deadline_format_helper.rb [hunk 6]
```diff
@@ -0,0 +1,8 @@
+# frozen_string_literal: true
+
+# Converts from Deadline to string representing datetime in correct format for forms
+module DeadlineFormatHelper
+  def self.convert(deadline)
+    deadline.deadline_at.strftime('%m/%d/%Y %H:%M') if deadline
```
Comment (nwoodthorpe) [inline app/helpers/deadline_format_helper.rb]: return ''

File: app/jobs/deadline_job.rb [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+  def perform(deadline_id)
+    deadline = Deadline.find_by(id: deadline_id)
+    return unless deadline
+
+    deadline.assignment.repos.each do |repo|
+      fetch_submission_sha(repo)
+    end
+  end
+
+  def fetch_submission_sha(repo)
```
Comment (nwoodthorpe) [inline app/jobs/deadline_job.rb]: Look into rate limit issues here

File: app/models/assignment/editor.rb [hunk 8]
```diff
@@ -61,6 +63,22 @@ def perform
 
     private
 
+    def handle_deadline(deadline)
```
Comment (nwoodthorpe) [inline app/models/assignment/editor.rb]: recreate deadline, move return if to perform

File: app/models/concerns/repo_submittable.rb [hunk 9]
```diff
@@ -0,0 +1,23 @@
+# frozen_string_literal: true
+
+module RepoSubmittable
```
Comment (nwoodthorpe) [inline app/models/concerns/repo_submittable.rb]: View model

File: app/models/assignment_repo.rb [hunk 10]
```diff
@@ -19,6 +21,8 @@ class AssignmentRepo < ApplicationRecord
 
   delegate :creator, :starter_code_repo_id, to: :assignment
   delegate :github_user,                    to: :user
+  delegate :default_branch,                 to: :github_repository
+  delegate :commits,                        to: :github_repository
```
Comment (nwoodthorpe) [inline app/models/assignment_repo.rb]: Merge this into line above

File: app/models/deadline.rb [hunk 11]
```diff
@@ -0,0 +1,37 @@
+# frozen_string_literal: true
+
+class Deadline < ApplicationRecord
+  belongs_to :assignment, polymorphic: true
+  validates :deadline_at, presence: true
+  validate :deadline_in_future
+
+  after_create :create_job
```
Comment (nwoodthorpe) [inline app/models/deadline.rb]: Move to factory

Comment (nwoodthorpe) [inline app/models/deadline.rb]: See if we can move job creation to editor/creator

File: app/models/group_assignment_repo.rb [hunk 12]
```diff
@@ -34,6 +36,8 @@ class GroupAssignmentRepo < ApplicationRecord
 
   delegate :creator, :starter_code_repo_id, to: :group_assignment
   delegate :github_team_id,                 to: :group
+  delegate :default_branch,                 to: :github_repository
+  delegate :commits,                        to: :github_repository
```
Comment (nwoodthorpe) [inline app/models/group_assignment_repo.rb]: Move to 39 ^

File: app/views/shared/_submission.html.erb [hunk 13]
```diff
@@ -0,0 +1,9 @@
+<% if repo.submission_failed? %>
+  <a class="submission-icon-container tooltipped tooltipped-n" aria-label="No Submission" href="#">
```
Comment (nwoodthorpe) [inline app/views/shared/_submission.html.erb]: SPAN

File: config/sidekiq.yml [hunk 14]
```diff
@@ -7,3 +7,4 @@
   - [last_active,  1]
```
Comment (nwoodthorpe) [inline config/sidekiq.yml]: Add link to documentation https://github.com/mperham/sidekiq/wiki/Advanced-Options#queues

File: db/schema.rb [hunk 15]
```diff
@@ -10,7 +10,7 @@
 #
 # It's strongly recommended that you check this file into your version control system.
 
-ActiveRecord::Schema.define(version: 20170319181810) do
```
Comment (nwoodthorpe) [inline db/schema.rb]: Re-run db:migrate with rails 5.1

File: db/migrate/20170526212851_create_deadlines_table.rb [hunk 16]
```diff
@@ -0,0 +1,11 @@
+class CreateDeadlinesTable < ActiveRecord::Migration[5.0]
+  def change
+    create_table :deadlines do |t|
+      t.string :name
+      t.references :assignment, polymorphic: true
+      t.datetime :deadline_at
```
Comment (nwoodthorpe) [inline db/migrate/20170526212851_create_deadlines_table.rb]: add required

File: db/migrate/20170526212851_create_deadlines_table.rb [hunk 17]
```diff
@@ -0,0 +1,11 @@
+class CreateDeadlinesTable < ActiveRecord::Migration[5.0]
+  def change
+    create_table :deadlines do |t|
+      t.string :name
```
Comment (nwoodthorpe) [inline db/migrate/20170526212851_create_deadlines_table.rb]: Remove name

**Comment by nwoodthorpe** [comment]
@tarebyte instead of creating creators for assignments and assignment repos, I've moved the job creation from ActiveModel callbacks to inside the controller after we save the assignment. I think we should still do the creators, but this PR is getting a bit massive so I'd rather do that separately.

**Comment by nwoodthorpe** [comment]
@tarebyte Do you see any issues coming up from changing `GitHub::Errors#raise_github_too_many_requests` to return a unique error class? It's raising GitHub::Error right now so the only way to check for rate limiting issues in the job would be to make another query to a [rate_limit api endpoint](https://developer.github.com/v3/rate_limit/) if we hit an error to figure out whether that was a rate limiting error.

**Comment by tarebyte** [comment]
@nwoodthorpe this is looking really good, can I convince you to extract _just_ the backend model bits to another PR so that I can review is in pieces?

This PR is getting a little large to review everything at once.

**Comment by nwoodthorpe** [comment]
@tarebyte Models are merged and the conflicts here are fixed :+1:

**Comment by nwoodthorpe** [comment]
🎆

**Comment by nwoodthorpe** [comment]
@tarebyte the newest commit adds a couple lines of logging so we can hopefully figure out what's up.

**Comment by nwoodthorpe** [comment]
Tested in production, both Assignment  and GroupAssignment deadlines work as expected.

@tarebyte this is good to merge!

**Pull request merged**